File: verify.py
#!/usr/bin/env python3
"""
verify.py — Reference verifier for ADP-1 (Akaeon Declaration Protocol v1).

This is the "you don't have to trust Akaeon" property, expressed as code.
It implements the normative verification procedure from ADP-1 §7 in full:

    1. Obtain the declaration record.
    2. Recompute the canonical (RFC 8785 / JCS) bytes and their SHA-256.
    3. Verify the declarant's signature over the canonical message.
    4. Verify the declarant's authority over the subject.
    5. Fetch the anchored Merkle root (from Arweave, NOT from Akaeon).
    6. Verify the Merkle inclusion proof reconstructs that root.
    7. Read the anchor's timestamp.

No Akaeon SDK. No proprietary cryptography. The only third party trusted is
Arweave (for permanence and block time). If every Akaeon server is gone, this
still verifies any declaration against the public Arweave network.

Dependencies: the `cryptography` package for Ed25519 (stdlib has no Ed25519
verify on all versions). Everything else is stdlib. If you do not want the
`cryptography` dependency, the 6-line pure-Python Ed25519 verify in §7.4 of the
spec drops in here unchanged.

Usage:
    # Offline, against the shipped test vectors (no network):
    python3 verify.py --vectors test-vectors.json

    # Online, against a live registry + Arweave gateway:
    python3 verify.py --declaration-id 01J... \\
        --registry https://api.akaeon-registry.com \\
        --gateway https://arweave.net
"""

import argparse
import base64
import hashlib
import json
import sys
import urllib.request

try:
    from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
    from cryptography.exceptions import InvalidSignature
except ImportError:  # pragma: no cover
    sys.stderr.write(
        "This reference verifier uses the `cryptography` package for Ed25519.\n"
        "Install it with: pip install cryptography\n"
        "(Or drop in the pure-Python Ed25519 verify from ADP-1 spec §7.4.)\n"
    )
    raise


# --------------------------------------------------------------------------- #
# Primitives
# --------------------------------------------------------------------------- #

def sha256(data: bytes) -> bytes:
    return hashlib.sha256(data).digest()


def sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def jcs_canonicalize(value) -> bytes:
    """RFC 8785 JSON Canonicalization Scheme, restricted to the ADP-1 field set.

    ADP-1 §4.7 constrains canonical declarations to: sorted object keys
    (by UTF-16 code unit, == codepoint order for the ASCII keys ADP-1 uses),
    ordered arrays, JSON-grammar string escaping, and integer-only numbers.
    Within those constraints Python's json.dumps with sort_keys and the
    compact separators is byte-identical to full JCS.
    """
    def check_numbers(v):
        if isinstance(v, bool):
            return
        if isinstance(v, float):
            raise ValueError("ADP-1 canonical fields forbid non-integer numbers")
        if isinstance(v, dict):
            for x in v.values():
                check_numbers(x)
        elif isinstance(v, list):
            for x in v:
                check_numbers(x)

    check_numbers(value)
    return json.dumps(
        value,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
    ).encode("utf-8")


def ed25519_verify(raw_pubkey_b64: str, message: bytes, signature_b64: str) -> bool:
    """Verify an Ed25519 signature given a raw 32-byte public key (base64)."""
    pub = Ed25519PublicKey.from_public_bytes(base64.b64decode(raw_pubkey_b64))
    try:
        pub.verify(base64.b64decode(signature_b64), message)
        return True
    except InvalidSignature:
        return False


def verify_inclusion(leaf_hash_hex, leaf_index, tree_size, proof_hex, root_hex) -> bool:
    """RFC 6962 §2.1.2 inclusion-proof verification.

    `tree_size` is load-bearing: RFC 6962 promotes (does not duplicate) the
    last node at an odd-count level, so whether the current node sits on the
    right edge depends on tree_size, not just the index's low bit. A
    Bitcoin-style verifier that ignores tree_size silently mis-hashes any
    path through a promoted node.
    """
    if leaf_index < 0 or leaf_index >= tree_size or tree_size == 0:
        return False
    fn = leaf_index
    sn = tree_size - 1
    r = bytes.fromhex(leaf_hash_hex)
    for p_hex in proof_hex:
        p = bytes.fromhex(p_hex)
        if sn == 0:
            return False  # proof longer than path
        if (fn & 1) == 1 or fn == sn:
            r = sha256(b"\x01" + p + r)
            while (fn & 1) == 0:
                fn >>= 1
                sn >>= 1
        else:
            r = sha256(b"\x01" + r + p)
        fn >>= 1
        sn >>= 1
    return sn == 0 and r.hex() == root_hex


# --------------------------------------------------------------------------- #
# The seven-step verification procedure (ADP-1 §7)
# --------------------------------------------------------------------------- #

class Result:
    def __init__(self):
        self.checks = []
        self.ok = True

    def add(self, name, passed, detail=""):
        self.checks.append((name, passed, detail))
        if not passed:
            self.ok = False

    def report(self):
        for name, passed, detail in self.checks:
            mark = "PASS" if passed else "FAIL"
            line = f"  [{mark}] {name}"
            if detail:
                line += f" — {detail}"
            print(line)
        print()
        print("  RESULT:", "VERIFIED" if self.ok else "VERIFICATION FAILED")


def verify_declaration(record, anchored_payload, authority_check=None) -> Result:
    """Verify one ADP-1 declaration bundle.

    `record` is the verification bundle for a single declaration (the shape in
    test-vectors.json `declarations[i]`, or the registry's
    /v1/public/declarations/<id>/verify response).

    `anchored_payload` is the batch payload fetched from Arweave (NOT from
    Akaeon). `authority_check`, if provided, is a callable
    (declaration) -> (bool, detail) that proves the declarant's standing over
    the subject (DNS lookup for domain scope, etc.). When None, the structural
    binding is checked but the live network proof is reported as DEFERRED.
    """
    res = Result()
    decl = record["declaration"]
    proof = record["declarant_proof"]

    # Step 2 — recompute canonical bytes + hash.
    canonical = jcs_canonicalize(decl)
    canonical_hash = sha256_hex(canonical)
    res.add(
        "2. Canonical hash matches record",
        canonical_hash == record["canonical_declaration_sha256"],
        f"sha256={canonical_hash[:16]}…",
    )

    # Step 3 — verify the declarant signature over the canonical message.
    expected_msg = f"adp:declaration:v1|{decl['declaration_id']}|{canonical_hash}"
    msg_ok = proof["canonical_message"] == expected_msg
    sig_ok = ed25519_verify(
        proof["public_key"], proof["canonical_message"].encode("utf-8"), proof["signature"]
    )
    # Bind the signing key to the declared key — a signature by some other key
    # is worthless.
    key_bound = proof["public_key"] == decl["declarant"]["public_key"]
    res.add("3a. Canonical message well-formed", msg_ok)
    res.add("3b. Declarant signature valid", sig_ok)
    res.add("3c. Signing key == declared declarant key", key_bound)

    # Step 4 — authority binding.
    method = decl["authority"]["method"]
    if authority_check is not None:
        passed, detail = authority_check(decl)
        res.add(f"4. Declarant authority over subject ({method})", passed, detail)
    else:
        # Offline: we can confirm the binding is structurally present and, for
        # did:web, that the declarant id is consistent with the subject domain.
        structural = True
        detail = "structural only — live proof DEFERRED (no network)"
        if method == "dns-01" and decl["subject"].get("scope") == "domain":
            did = decl["declarant"]["id"]
            structural = did == f"did:web:{decl['subject']['domain']}"
            detail = f"did:web binds to subject domain ({'consistent' if structural else 'MISMATCH'}); live DNS DEFERRED"
        res.add(f"4. Declarant authority over subject ({method})", structural, detail)

    # Step 5 + 6 — anchored root, inclusion proof.
    anchored_root = anchored_payload["merkle_root_sha256"]
    root_consistent = anchored_root == record["merkle_root_sha256"]
    res.add("5. Anchored root == record's claimed root", root_consistent,
            f"root={anchored_root[:16]}…")

    incl = verify_inclusion(
        record["leaf_hash_sha256"],
        record["leaf_index"],
        record["tree_size"],
        record["merkle_proof"],
        anchored_root,
    )
    res.add("6. Merkle inclusion proof reconstructs anchored root", incl)

    # Also re-derive the leaf hash from the canonical bytes (strongest posture):
    leaf_recomputed = sha256(b"\x00" + canonical).hex()
    res.add("6b. Leaf hash == SHA-256(0x00 || canonical bytes)",
            leaf_recomputed == record["leaf_hash_sha256"])

    # The anchor operator's signature over the root (binds root↔batch key).
    anchor_sig = anchored_payload["anchor_signature"]
    anchor_msg = f"adp:anchor:v1|{anchored_payload['batch_id']}|{anchored_root}|{anchored_payload['leaf_count']}"
    anchor_sig_ok = (
        anchor_sig["canonical_message"] == anchor_msg
        and ed25519_verify(anchor_sig["public_key"],
                            anchor_sig["canonical_message"].encode("utf-8"),
                            anchor_sig["signature"])
    )
    res.add("5b. Anchor signature over root valid", anchor_sig_ok)

    return res


# --------------------------------------------------------------------------- #
# Drivers — offline (test vectors) and online (registry + gateway)
# --------------------------------------------------------------------------- #

def run_offline(vectors_path):
    with open(vectors_path, "r", encoding="utf-8") as f:
        vectors = json.load(f)
    anchored = vectors["anchor"]["anchored_payload"]
    overall_ok = True
    for i, rec in enumerate(vectors["declarations"]):
        print(f"Declaration [{i}] — {rec['description']}")
        print(f"  id: {rec['declaration']['declaration_id']}")
        res = verify_declaration(rec, anchored, authority_check=None)
        res.report()
        print()
        overall_ok = overall_ok and res.ok
    return overall_ok


def http_get_json(url):
    with urllib.request.urlopen(url, timeout=30) as r:
        return json.loads(r.read().decode("utf-8"))


def run_online(declaration_id, registry, gateway):
    record = http_get_json(f"{registry}/v1/public/declarations/{declaration_id}/verify")
    # Fetch the anchored payload from Arweave — NOT from the registry.
    tx_id = record["anchor"]["arweave_tx_id"]
    anchored = http_get_json(f"{gateway}/{tx_id}")

    def dns_authority(decl):
        # Live DNS authority proof for domain scope. Kept dependency-free:
        # uses the gateway's resolver via a DoH query would go here. For the
        # reference we report the structural binding and direct the operator to
        # the spec §6.1 DNS procedure for the production resolver step.
        if decl["authority"]["method"] == "dns-01":
            return (True, "DNS authority: run multi-resolver TXT check per spec §6.1")
        return (True, f"authority method {decl['authority']['method']}: see spec §5.6")

    print(f"Declaration — {declaration_id}")
    res = verify_declaration(record, anchored, authority_check=dns_authority)
    res.report()
    return res.ok


def main():
    ap = argparse.ArgumentParser(description="ADP-1 reference verifier")
    ap.add_argument("--vectors", help="path to test-vectors.json (offline mode)")
    ap.add_argument("--declaration-id", help="declaration id to verify (online mode)")
    ap.add_argument("--registry", default="https://api.akaeon-registry.com")
    ap.add_argument("--gateway", default="https://arweave.net")
    args = ap.parse_args()

    if args.vectors:
        ok = run_offline(args.vectors)
    elif args.declaration_id:
        ok = run_online(args.declaration_id, args.registry, args.gateway)
    else:
        ap.error("supply --vectors (offline) or --declaration-id (online)")

    sys.exit(0 if ok else 1)


if __name__ == "__main__":
    main()

File: verify.mjs
#!/usr/bin/env node
// verify.mjs — Reference verifier for ADP-1 (Akaeon Declaration Protocol v1).
//
// Standard library only (node:crypto has Ed25519 built in — zero dependencies).
// Implements the normative seven-step procedure from ADP-1 §7:
//
//   1. Obtain the declaration record.
//   2. Recompute canonical (RFC 8785 / JCS) bytes + SHA-256.
//   3. Verify the declarant signature over the canonical message.
//   4. Verify the declarant's authority over the subject.
//   5. Fetch the anchored Merkle root from Arweave (NOT from Akaeon).
//   6. Verify the Merkle inclusion proof reconstructs that root.
//   7. Read the anchor timestamp.
//
// No Akaeon SDK, no proprietary cryptography, Akaeon trusted for nothing.
//
//   node verify.mjs --vectors test-vectors.json          # offline (no network)
//   node verify.mjs --vectors test-vectors.json --tamper  # prove it REJECTS forgeries
//   node verify.mjs --declaration-id 01J... \
//       --registry https://api.akaeon-registry.com --gateway https://arweave.net

import crypto from 'node:crypto'
import fs from 'node:fs'

const SPKI_PUB_PREFIX = Buffer.from('302a300506032b6570032100', 'hex')
const sha256 = (b) => crypto.createHash('sha256').update(b).digest()
const sha256hex = (b) => sha256(b).toString('hex')

// --- RFC 8785 JCS, restricted to the ADP-1 field set (spec §4.7). ----------
function jcsCanonicalize(value) {
  return Buffer.from(ser(value), 'utf8')
  function ser(v) {
    if (v === null) return 'null'
    if (typeof v === 'boolean') return v ? 'true' : 'false'
    if (typeof v === 'number') {
      if (!Number.isFinite(v)) throw new Error('non-finite number')
      if (!Number.isInteger(v)) throw new Error('ADP-1 canonical fields forbid non-integer numbers')
      return String(v)
    }
    if (typeof v === 'string') return JSON.stringify(v)
    if (Array.isArray(v)) return '[' + v.map(ser).join(',') + ']'
    if (typeof v === 'object') {
      const keys = Object.keys(v).sort()
      const parts = []
      for (const k of keys) {
        if (v[k] === undefined) continue
        parts.push(JSON.stringify(k) + ':' + ser(v[k]))
      }
      return '{' + parts.join(',') + '}'
    }
    throw new Error('unserializable: ' + typeof v)
  }
}

// --- Ed25519 verify from a raw 32-byte public key (base64). ----------------
function ed25519Verify(rawPubB64, message, signatureB64) {
  const pub = crypto.createPublicKey({
    key: Buffer.concat([SPKI_PUB_PREFIX, Buffer.from(rawPubB64, 'base64')]),
    format: 'der',
    type: 'spki',
  })
  return crypto.verify(null, Buffer.from(message, 'utf8'), pub, Buffer.from(signatureB64, 'base64'))
}

// --- RFC 6962 §2.1.2 inclusion-proof verification. -------------------------
function verifyInclusion(leafHashHex, leafIndex, treeSize, proofHex, rootHex) {
  if (leafIndex < 0 || leafIndex >= treeSize || treeSize === 0) return false
  let fn = leafIndex
  let sn = treeSize - 1
  let r = Buffer.from(leafHashHex, 'hex')
  for (const pHex of proofHex) {
    const p = Buffer.from(pHex, 'hex')
    if (sn === 0) return false
    if ((fn & 1) === 1 || fn === sn) {
      r = sha256(Buffer.concat([Buffer.from([0x01]), p, r]))
      while ((fn & 1) === 0) { fn >>= 1; sn >>= 1 }
    } else {
      r = sha256(Buffer.concat([Buffer.from([0x01]), r, p]))
    }
    fn >>= 1
    sn >>= 1
  }
  return sn === 0 && r.toString('hex') === rootHex
}

// --- The seven-step procedure. ---------------------------------------------
function verifyDeclaration(record, anchored, authorityCheck) {
  const checks = []
  const add = (name, passed, detail = '') => checks.push({ name, passed, detail })
  const decl = record.declaration
  const proof = record.declarant_proof

  // Step 2 — canonical bytes + hash.
  const canonical = jcsCanonicalize(decl)
  const canonicalHash = sha256hex(canonical)
  add('2. Canonical hash matches record', canonicalHash === record.canonical_declaration_sha256,
      `sha256=${canonicalHash.slice(0, 16)}…`)

  // Step 3 — declarant signature.
  const expectedMsg = `adp:declaration:v1|${decl.declaration_id}|${canonicalHash}`
  add('3a. Canonical message well-formed', proof.canonical_message === expectedMsg)
  add('3b. Declarant signature valid',
      ed25519Verify(proof.public_key, proof.canonical_message, proof.signature))
  add('3c. Signing key == declared declarant key',
      proof.public_key === decl.declarant.public_key)

  // Step 4 — authority binding.
  const method = decl.authority.method
  if (authorityCheck) {
    const { passed, detail } = authorityCheck(decl)
    add(`4. Declarant authority over subject (${method})`, passed, detail)
  } else {
    let structural = true
    let detail = 'structural only — live proof DEFERRED (no network)'
    if (method === 'dns-01' && decl.subject.scope === 'domain') {
      structural = decl.declarant.id === `did:web:${decl.subject.domain}`
      detail = `did:web binds to subject domain (${structural ? 'consistent' : 'MISMATCH'}); live DNS DEFERRED`
    }
    add(`4. Declarant authority over subject (${method})`, structural, detail)
  }

  // Steps 5 + 6 — anchored root + inclusion proof.
  const anchoredRoot = anchored.merkle_root_sha256
  add('5. Anchored root == record\'s claimed root', anchoredRoot === record.merkle_root_sha256,
      `root=${anchoredRoot.slice(0, 16)}…`)
  add('6. Merkle inclusion proof reconstructs anchored root',
      verifyInclusion(record.leaf_hash_sha256, record.leaf_index, record.tree_size,
                      record.merkle_proof, anchoredRoot))
  add('6b. Leaf hash == SHA-256(0x00 || canonical bytes)',
      sha256(Buffer.concat([Buffer.from([0x00]), canonical])).toString('hex') === record.leaf_hash_sha256)

  // Anchor operator's signature over the root.
  const as_ = anchored.anchor_signature
  const anchorMsg = `adp:anchor:v1|${anchored.batch_id}|${anchoredRoot}|${anchored.leaf_count}`
  add('5b. Anchor signature over root valid',
      as_.canonical_message === anchorMsg &&
      ed25519Verify(as_.public_key, as_.canonical_message, as_.signature))

  const ok = checks.every((c) => c.passed)
  return { checks, ok }
}

function report(label, decl, result) {
  console.log(label)
  if (decl) console.log(`  id: ${decl.declaration_id}`)
  for (const { name, passed, detail } of result.checks) {
    console.log(`  [${passed ? 'PASS' : 'FAIL'}] ${name}${detail ? ` — ${detail}` : ''}`)
  }
  console.log(`\n  RESULT: ${result.ok ? 'VERIFIED' : 'VERIFICATION FAILED'}\n`)
}

// --- Drivers. --------------------------------------------------------------
function runOffline(vectorsPath, tamper) {
  const vectors = JSON.parse(fs.readFileSync(vectorsPath, 'utf8'))
  const anchored = vectors.anchor.anchored_payload
  let allOk = true

  for (let i = 0; i < vectors.declarations.length; i++) {
    const rec = JSON.parse(JSON.stringify(vectors.declarations[i])) // deep copy
    if (tamper) {
      // Flip one field of the declaration AFTER signing. A correct verifier
      // must reject: the recomputed canonical hash no longer matches the
      // signed canonical message, and the leaf no longer matches the tree.
      rec.declaration.effective_from = '1999-01-01T00:00:00Z'
    }
    const res = verifyDeclaration(rec, anchored, null)
    report(`Declaration [${i}] — ${rec.description}${tamper ? '  [TAMPERED effective_from]' : ''}`,
           rec.declaration, res)
    if (tamper) {
      // In tamper mode, PASS means the verifier correctly REJECTED.
      const rejected = !res.ok
      console.log(`  [${rejected ? 'PASS' : 'FAIL'}] verifier rejected the tampered record\n`)
      allOk = allOk && rejected
    } else {
      allOk = allOk && res.ok
    }
  }
  return allOk
}

async function runOnline(declarationId, registry, gateway) {
  const record = await fetch(`${registry}/v1/public/declarations/${declarationId}/verify`).then((r) => r.json())
  const txId = record.anchor.arweave_tx_id
  const anchored = await fetch(`${gateway}/${txId}`).then((r) => r.json())
  const authorityCheck = (decl) =>
    decl.authority.method === 'dns-01'
      ? { passed: true, detail: 'DNS authority: run multi-resolver TXT check per spec §6.1' }
      : { passed: true, detail: `authority method ${decl.authority.method}: see spec §5.6` }
  const res = verifyDeclaration(record, anchored, authorityCheck)
  report(`Declaration — ${declarationId}`, record.declaration, res)
  return res.ok
}

// --- CLI. ------------------------------------------------------------------
const args = process.argv.slice(2)
const getArg = (name) => {
  const i = args.indexOf(name)
  return i >= 0 ? args[i + 1] : undefined
}

const vectors = getArg('--vectors')
const declarationId = getArg('--declaration-id')
const tamper = args.includes('--tamper')

let ok
if (vectors) {
  ok = runOffline(vectors, tamper)
} else if (declarationId) {
  ok = await runOnline(declarationId, getArg('--registry') || 'https://api.akaeon-registry.com',
                       getArg('--gateway') || 'https://arweave.net')
} else {
  console.error('supply --vectors (offline) or --declaration-id (online)')
  process.exit(2)
}
process.exit(ok ? 0 : 1)
